### PR TITLE
Implement automatic installation of LilyPond

### LINKED ISSUE #313: Idea: automatic installation/update of LilyPond
This post http://lists.gnu.org/archive/html/lilypond-user/2013-12/msg00238.html
made me think if it wouldn't be nice to add a function to download, install and configure LilyPond.

I'd say that on startup Frescobaldi checks for LilyPond and suggests to install if none is present. Later this would be only on request (i.e. menu item).

It should be configurable if only stable or also development releases should be looked for.

I can try it if it seems worthwile.

### PR DESCRIPTION
**Pull request by jeanas**
This needs as much testing as it can! In particular, I have not tested it under macOS or Windows.

Fixes #313

### CODE REVIEW

**Comment by karlinhigh** [comment]
> testing ... Windows.

I am very glad to see this effort. I am trying to test this on Windows, following the [instructions ](https://github.com/frescobaldi/frescobaldi/blob/master/INSTALL.md)for installing from source.

I have Windows 11 21H2 and Python 3.11.4 installed from Microsoft Store, build requirements installed via pip. It is giving build errors:

<details>
<summary>tox -e mo-generate</summary>
<code>
D:\FrescobaldiTest\frescobaldi>tox -e mo-generate
mo-generate: commands[0]> python i18n/mo-gen.py
Traceback (most recent call last):
  File "D:\FrescobaldiTest\frescobaldi\i18n\mo-gen.py", line 16, in <module>
    compile_mo(po, "frescobaldi")
  File "D:\FrescobaldiTest\frescobaldi\i18n\mo-gen.py", line 11, in compile_mo
    subprocess.run(["msgfmt", "-o", out_file, path])
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\Lib\subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\Lib\subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\Lib\subprocess.py", line 1538, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [WinError 2] The system cannot find the file specified
mo-generate: exit 1 (0.70 seconds) D:\FrescobaldiTest\frescobaldi> python i18n/mo-gen.py pid=16420
  mo-generate: FAIL code 1 (5.86=setup[5.16]+cmd[0.70] seconds)
  evaluation failed :( (6.52 seconds)
</code>
</details>

<details>
<summary>pipx install --system-site-packages --editable ./frescobaldi</summary>
Fatal error from pip prevented installation. Full pip output in file:
    C:\Users\owner\.local\pipx\logs\cmd_2023-08-03_17.59.53_pip_errors.log

pip seemed to fail to build package:
    python-poppler-qt5

Some possibly relevant errors from pip install:
    error: subprocess-exited-with-error
    error: metadata-generation-failed

Error installing frescobaldi from spec 'D:\\FrescobaldiTest\\frescobaldi'.
</details>

<details>
<summary>Contents of C:\Users\owner\.local\pipx\logs\cmd_2023-08-03_17.59.53_pip_errors.log</summary>
<code>
PIP STDOUT
 ----------
Obtaining file:///D:/FrescobaldiTest/frescobaldi
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Checking if build backend supports build_editable: started
  Checking if build backend supports build_editable: finished with status 'done'
  Getting requirements to build editable: started
  Getting requirements to build editable: finished with status 'done'
  Installing backend dependencies: started
  Installing backend dependencies: finished with status 'done'
  Preparing editable metadata (pyproject.toml): started
  Preparing editable metadata (pyproject.toml): finished with status 'done'
Requirement already satisfied: pyqt5>=5.9 in c:\users\owner\appdata\local\packages\pythonsoftwarefoundation.python.3.11_qbz5n2kfra8p0\localcache\local-packages\python311\site-packages (from frescobaldi==3.3.0) (5.15.9)
Collecting python-ly>=0.9.5 (from frescobaldi==3.3.0)
  Downloading python-ly-0.9.7.tar.gz (183 kB)
     -------------------------------------- 184.0/184.0 kB 5.6 MB/s eta 0:00:00
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Collecting python-poppler-qt5 (from frescobaldi==3.3.0)
  Using cached python-poppler-qt5-21.3.0.tar.gz (28 kB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'error'
</code>
<code>PIP STDERR
----------
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [23 lines of output]
      Traceback (most recent call last):
        File "C:\Users\owner\.local\pipx\shared\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
          main()
        File "C:\Users\owner\.local\pipx\shared\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\owner\.local\pipx\shared\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 152, in prepare_metadata_for_build_wheel
          whl_basename = backend.build_wheel(metadata_directory, config_settings)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\owner\AppData\Local\Temp\pip-build-env-grof3y8y\overlay\Lib\site-packages\sipbuild\api.py", line 46, in build_wheel
          project = AbstractProject.bootstrap('wheel',
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\owner\AppData\Local\Temp\pip-build-env-grof3y8y\overlay\Lib\site-packages\sipbuild\abstract_project.py", line 87, in bootstrap
          project.setup(pyproject, tool, tool_description)
        File "C:\Users\owner\AppData\Local\Temp\pip-build-env-grof3y8y\overlay\Lib\site-packages\sipbuild\project.py", line 586, in setup
          self.apply_user_defaults(tool)
        File "C:\Users\owner\AppData\Local\Temp\pip-build-env-grof3y8y\overlay\Lib\site-packages\pyqtbuild\project.py", line 70, in apply_user_defaults
          super().apply_user_defaults(tool)
        File "C:\Users\owner\AppData\Local\Temp\pip-build-env-grof3y8y\overlay\Lib\site-packages\sipbuild\project.py", line 237, in apply_user_defaults
          self.builder.apply_user_defaults(tool)
        File "C:\Users\owner\AppData\Local\Temp\pip-build-env-grof3y8y\overlay\Lib\site-packages\pyqtbuild\builder.py", line 69, in apply_user_defaults
          raise PyProjectOptionException('qmake',
      sipbuild.pyproject.PyProjectOptionException
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

</code>
</details>

**Comment by karlinhigh** [comment]
> just download the Frescobaldi source code and drop the `frescobaldi_app` directory into an existing Frescobaldi installation obtained with the Frescobaldi 3.3 installer.

That sounds much easier. I installed Frescobaldi 3.3.0 and accidentally confirmed it runs: I was not working off the correct GitHub repository, and it didn't have the code from this pull request.

Then I got the GitHub CLI set up and did `gh repo clone frescobaldi/frescobaldi` and then `gh pr checkout 1655`. I confirmed the new code is present in `C:\Program Files (x86)\Frescobaldi\lib\frescobaldi_app`. Upon starting Frescobaldi, it gave this message box and closed:

image

Disclosure: I didn't yet restart the computer like the Frescobaldi installer wanted.

**Comment by karlinhigh** [comment]
Wait, how about if I just put in the 4 files affected by this PR, rather than the whole folder? The "stock" folder has everything in `.pyc` files apparently. Progress!

image

Downloads get stuck at 100% for some reason. 

Based on this code...
```
LILYPOND_AUTOINSTALL_DIR = (
    os.path.join(QStandardPaths.writableLocation(QStandardPaths.DataLocation),
                 "lilypond-binaries")
)
```
...I searched for `lilypond-binaries` and found this directory, which remains empty:

`C:\Users\owner\AppData\Local\frescobaldi\frescobaldi\lilypond-binaries`

Running Frescobaldi as Administrator didn't make a difference.

**Comment by karlinhigh** [comment]
Windows' CMD shell doesn't wait for Frescobaldi to exit, so I switched to Git Bash instead. Here's what it said:
```
owner@FRONTDESK MINGW64 /c/Program Files (x86)/Frescobaldi
$ ./frescobaldi.exe
Traceback (most recent call last):
  File "C:\Program Files (x86)\Frescobaldi\lib\frescobaldi_app\preferences\lilypond.py", line 339, in downloadFinished
    shutil.unpack_archive(tfile.name, lilypondinfo.LILYPOND_AUTOINSTALL_DIR,
  File "C:\Python\Python310-32\lib\shutil.py", line 1262, in unpack_archive
  File "C:\Python\Python310-32\lib\shutil.py", line 1170, in _unpack_zipfile
shutil.ReadError: C:\Users\owner\AppData\Local\Temp\tmpk9zbu_1q is not a zip file
Error in sys.excepthook:
Traceback (most recent call last):
  File "D:\Users\Jones\Projekte\Python\frescobaldi\windows\..\frescobaldi_app\app.py", line 266, in excepthook
  File "D:\Users\Jones\Projekte\Python\frescobaldi\windows\..\frescobaldi_app\exception.py", line 55, in __init__
  File "D:\Users\Jones\Projekte\Python\frescobaldi\windows\..\frescobaldi_app\extensions\__init__.py", line 801, in is_extension_exception
  File "C:\Python\Python310-32\lib\re.py", line 251, in compile
  File "C:\Python\Python310-32\lib\re.py", line 303, in _compile
  File "C:\Python\Python310-32\lib\sre_compile.py", line 764, in compile
  File "C:\Python\Python310-32\lib\sre_parse.py", line 964, in parse
re.error: unbalanced parenthesis at position 15

Original exception was:
Traceback (most recent call last):
  File "C:\Program Files (x86)\Frescobaldi\lib\frescobaldi_app\preferences\lilypond.py", line 339, in downloadFinished
    shutil.unpack_archive(tfile.name, lilypondinfo.LILYPOND_AUTOINSTALL_DIR,
  File "C:\Python\Python310-32\lib\shutil.py", line 1262, in unpack_archive
  File "C:\Python\Python310-32\lib\shutil.py", line 1170, in _unpack_zipfile
shutil.ReadError: C:\Users\owner\AppData\Local\Temp\tmpk9zbu_1q is not a zip file
```
It gives file paths that do not exist on my machine. I am guessing that comes from development or build environments.

**Comment by karlinhigh** [comment]
GitHub: "File size too big: 25 MB are allowed, 37 MB were attempted to upload."

But I can confirm that the `temp-archive.zip` file downloaded to the home directory does contain a working LilyPond 2.24.1.

However, that test was run on a different computer than earlier today. This one has Windows 11 22H2.

Naturally, I then had to try the `lilypond.py` file from the original commit. And that one worked as well on this second computer. It now has 4 different LilyPonds under `AppData\Local\frescobaldi\frescobaldi\lilypond-binaries`, namely 2.24.1 and 2.25.5 through .7, all placed there by Frescobaldi. Some were simultaneous downloads. And the option for "Automatically choose LilyPond version from document" is doing the right thing for each of them in response to changing the `\version` statement.

That is as far as I can go for today. I do plan to later re-visit the first computer where this wasn't working.

**Comment by karlinhigh** [comment]
> I do plan to later re-visit the first computer where this wasn't working.

* Uninstalled Microsoft Store's Python 3.11
* Re-installed Frescobaldi
* Now automatic installation works.

🤷‍♂️

Next, on to macOS.

**Comment by karlinhigh** [comment]
Testing with: MacBook Air (11-inch, Mid 2012), Intel Core i5-3317U, macOS Catalina 10.15.7

Downloaded Frescobaldi 3.3.0's DMG, installed as usual, gave permission to it in System Preferences -> Security & Privacy.

It took me a little to find the `frescobaldi_app` location: `/Applications/Frescobaldi.app/Contents/Resources/lib/python3.10/frescobaldi_app/`

Automatic installation of LilyPond worked. I got 2.24.1, and 2.25.5 though .7 with almost no problems. There was a dialog saying something like "Running LilyPond, this could take a while" which I eventually cancelled out of. Automatic version selection is working, none of the versions needed any further permissions to run.

**Comment by karlinhigh** [comment]
On Tue, Aug 8, 2023, 6:30 PM Jean Abou-Samra ***@***.***>
wrote:

> dialog: did you wait a few dozens of seconds or a few minutes?
>
The dialog appeared before I had closed the Preferences window. I wasn't
aware of having started any of the executables. I gave it about a minute,
I'd say.

>

File: frescobaldi_app/preferences/lilypond.py [hunk 1]
```diff
@@ -173,24 +230,293 @@ def itemChanged(self, item):
...
+        else:
+            raise AssertionError
 
-    def display(self):
-        text = self._info.prettyName()
-        if self._info.version():
-            self.setIcon(icons.get("lilypond-run"))
+        # Download an archive of LilyPond binaries from GitLab
+        archive_url = QUrl(
+            "https://gitlab.com/api/v4/projects/lilypond%2Flilypond/packages/"
```
Comment (PeterBjuhr) [inline frescobaldi_app/preferences/lilypond.py:303]: suggestion: Use the url as a constant.

File: frescobaldi_app/preferences/lilypond.py [hunk 2]
```diff
@@ -173,24 +230,293 @@ def itemChanged(self, item):
...
+            exec_name = "lilypond.exe"
         else:
-            self.setIcon(icons.get("dialog-error"))
-        if self._info.command == self.listWidget().parentWidget().parentWidget()._defaultCommand:
-            text += " [{0}]".format(_("default"))
+            exec_name = "lilypond"
+        self.executable_path = os.path.join(lilypondinfo.LILYPOND_AUTOINSTALL_DIR,
+                                            f"lilypond-{version_string}", "bin",
+                                            exec_name)
+
```
Comment (PeterBjuhr) [inline frescobaldi_app/preferences/lilypond.py:319]: nitpick: Remove this empty row.

File: frescobaldi_app/preferences/lilypond.py [hunk 3]
```diff
@@ -173,24 +230,293 @@ def itemChanged(self, item):
...
+        assert isinstance(self.state, DownloadingState)
+        self.state.download_progress = progress
+        self.display()
+
+    def downloadFinished(self):
+        if isinstance(self.state, DownloadErrorState):
+            return # the finished signal is called even in the error case
+        assert isinstance(self.state, DownloadingState)
+        # Unpack the downloaded archive
+        with open(os.path.expanduser("~/temp-archive.zip"), "wb") as tfile:
```
Comment (PeterBjuhr) [inline frescobaldi_app/preferences/lilypond.py:337]: suggestion: use this path as a constant.

File: frescobaldi_app/preferences/lilypond.py [hunk 4]
```diff
@@ -173,24 +230,293 @@ def itemChanged(self, item):
...
+        self.downloadSuccessful = True
+
+        # Decode the JSON data returned by the API
+        data = json.loads(bytes(self.reply.readAll()))
+        versions = []
+        for item in data:
+            tag = item["tag_name"]
+            # This release dates back to before LilyPond binaries became simple
+            # for us to install. (Its tag also uses a different naming convention.)
+            if tag == "release/2.22.2-1":
```
Comment (PeterBjuhr) [inline frescobaldi_app/preferences/lilypond.py:422]: question: How come we can work with specific versions here?

**Comment by PeterBjuhr** [comment]
Review done! :slightly_smiling_face:

**Comment by karlinhigh** [comment]
> try to diagnose if it ever reproduces for someone else.

I probably won't complete more testing in the next 2 days.

My first guess for the problems on the Windows computer would be a fight between the Pythons, the one from Microsoft Store and the one with Fresocbaldi.

On the macOS, I'd have to delete the app folder for Frescobaldi and start over to see if the problem happens again.

**Review by PeterBjuhr — APPROVED** [review]

**Pull request merged**